pkg/kubelet/apis/config/types.go

```diff
@@ -382,6 +385,12 @@ type KubeletConfiguration struct {
 	// Defaults to 10 seconds, requires GracefulNodeShutdown feature gate to be enabled.
 	// For example, if ShutdownGracePeriod=30s, and ShutdownGracePeriodCriticalPods=10s, during a node shutdown the first 20 seconds would be reserved for gracefully terminating normal pods, and the last 10 seconds would be reserved for terminating critical pods.
 	ShutdownGracePeriodCriticalPods metav1.Duration
+	// ReservedMemory specifies a comma separated list of memory reservations for NUMA nodes.
+	// Each configuration describes the reserved memory for desired memory types on a specific NUMA node.
+	// The Memory Manager validates whether total amount of reserved memory is identical to memory reserved by the Node Allocatable feature.
+	// The format is [{"numaNode": integer, "limits": {"type": "limit"}}, {"numaNode": integer, "limits": {"type": "limit"}}, ...]
+	// (e.g. [{"numaNode":integer, "limits":{"memory": "512Mi", "hugepages-2Mi": "512Mi"}}]
+	ReservedMemory []MemoryReservation
```

Comment:
How does this interact with the `SystemReserved`, `KubeReserved`, and `QOSReserved` flags? The documentation in the field and flag need to explain it.  Can you describe it for me here in a comment so i can review before it gets changed in the code?

Reply:
Hi @smarterclayton ,
 
Could you provide a reference for `QOSReserved`? (I encounter this term first time)

I hope the excerpt from the draft of official documentation helps (notice that the syntax for `--reserved-memory=` is outdated, we will update it in the final doc):

### Reserved memory flag

[Node Allocatable Feature](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/) is commonly used by node administrators to reserve K8S node system resources for the kubelet or operating system processes in order to enhance the node stability. A dedicated set of flags can be used for this purpose to set the total amount of reserved memory for a node. This pre-configured value is subsequently utilized to calculate the real amount of node's "allocatable" memory available to pods. Also, K8S scheduler incorporates "allocatable" to optimise pod scheduling process. The foregoing flags include `--kube-reserved`, `--system-reserved` and `--eviction-threshold`. The sum of their values will account for the total amount of reserved memory.    

A new `--reserved-memory` flag was added to Memory Manager to allow for this total reserved memory to be split (by a node administrator) and accordingly reserved across many NUMA nodes. 

Syntax: 

`--reserved-memory=[{numa-node=int,type=string,limit=string}][,][...]`
* `numa-node` index, e.g. `0`
* `type` of memory:
  * `memory` - conventional memory
  * `hugepages-2Mi` or `hugepages-1Gi` - hugepages 
* `limit` - the amount of reserved memory, e.g. `1Gi`

Example usage:

`--reserved-memory={numa-node=0,type=memory,limit=1Gi},{numa-node=1,type=memory,limit=2Gi}`

When you specify values for `--reserved-memory` flag, you must comply with the setting that you prior provided via Node Allocatable Feature flags. That is, the following rule must be obeyed for each memory type: 

`sum(reserved-memory(i)) = kube-reserved + system-reserved + eviction-threshold`, 

where `i` is an index of a NUMA node. 

If you do not follow the formula above, the Memory Manager will show an error on startup.

In other words, the example above illustrates that for the conventional memory (`type=memory`), we reserve `3Gi` in total, i.e.: 

`sum(reserved-memory(i)) = reserved-memory(0) + reserved-memory(1) = 1Gi + 2Gi = 3Gi`

An example of Node Allocatable Feature flags [configuration](https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/):
* `--kube-reserved=cpu=500m,memory=50Mi`
* `--system-reserved=cpu=123m,memory=333Mi`
* `--eviction-hard=memory.available<500Mi`

_NOTICE: hard eviction threshold is not equal to zero by default but `100Mi`, so do not forget to decrease the total amount set via `--reserved-memory` by this `100Mi`. Otherwise, the Memory Manager will display an error. Here is an example of a correct configuration:_

```
--feature-gates=MemoryManager=true 
--kube-reserved=cpu=4,memory=4Gi 
--system-reserved=cpu=1,memory=1Gi 
--memory-manager-policy=static 
--reserved-memory={numa-node=0,type=memory,limit=3Gi},{numa-node=1,type=memory,limit=2148Mi}
```
Let us validate the configuration above:
1. `kube-reserved + system-reserved + eviction-hard(default) = reserved-memory(0) + reserved-memory(1)`
2. `4Gi + 1Gi + 100Mi = 3Gi + 2148Mi`
3. `5120Mi + 100Mi = 3072Mi + 2148Mi`
4. `5220Mi = 5220Mi` (correct!)

Reply:
Sure, we expect that the following formula must hold for each memory type:
`sum(ReservedMemory(i)) = kube-reserved + system-reserved + hard-eviction-threshold`, it means that if
```
--kube-reserved=cpu=500m,memory=250Mi
--system-reserved=cpu=500m,memory=250Mi
--eviction-hard=memory.available<500Mi
250+250+500=1Gi
```
the total amount of memory defined under the reserved-memory flag should be equal to 1Gi, it can be different combinations, like, reserve all memory from the first NUMA node `--reserved-memory 0:memory=1Gi` or reserve equally from two NUMA nodes `--reserved-memory 0:memory=500Mi --reserved-memory 1:memory=500Mi`.

Until the documentation PR will be merged we can point to the KEP section, and once it merged I will update the link to point to the documentation(the same way that we have under the CLI flag).